src/lib/testing/mocks.ts

```diff
@@ -0,0 +1,81 @@
+import { SagaFunc } from '../types';
+
+export interface CallMock<Return> {
+  type: 'call';
+  used: boolean;
+  func: () => Return;
+  value: Return;
+}
+
+export interface SelectMock<State, Return> {
+  type: 'select';
+  used: boolean;
+  func: (state: State) => Return;
+  value: Return;
+}
+
+export interface RunMock<State, Return> {
+  type: 'run';
+  used: boolean;
+  func: SagaFunc<State, any[], Return>;
+  value: Return;
+}
+
+export interface SpawnMock<State, Return> {
+  type: 'spawn';
+  used: boolean;
+  func: SagaFunc<State, any[], Return>;
+  value: Return;
+}
+
+export type Mock<State> = CallMock<any> | SelectMock<State, any> | RunMock<State, any> | SpawnMock<State, any>;
+
+export type Mocks<State> = Array<Mock<State>>;
+
+export const getSelectMocks = <State>(mocks: Mocks<State>) =>
+  mocks.filter((mock): mock is SelectMock<State, any> => mock.type === 'select');
+
+export const getCallMocks = <State>(mocks: Mocks<State>) =>
+  mocks.filter((mock): mock is CallMock<any> => mock.type === 'call');
+
+export const getRunMocks = <State>(mocks: Mocks<State>) =>
+  mocks.filter((mock): mock is RunMock<State, any> => mock.type === 'run');
+
+export const getMocks = <State, MockType extends Mock<State>>(mocks: Mocks<State>, type: Mock<State>['type']) =>
+  mocks.filter((mock): mock is MockType => mock.type === type);
+
+export function call<Return>(func: () => Return, value: Return): CallMock<Return> {
```

Comment:
@HenriBeck This typing doesn't work well (since TypeScript 3.4?) anymore.

You could pass anything for `value` and it would use that to widen the type of `func`.

I rewrote it a little to then use `ReturnType<F>` which seems to work.

Do you think we also need to check the values provided to `func` for a match?

Reply:
As this is only a mock, there is currently no way of also checking the arguments and only providing the return value. If necessary, we could add a way of checking against arguments though we might need a different syntax for this then.

Reply:
@tp I just tried and for me the type checking works

Reply:
Nvm, I was still on 3.3 in my workspace

Reply:
Does this only concern `call` or is it also in `run` etc.? `select` seems to work fine

Reply:
@HenriBeck hmm, I was able to use `call(sleep, null)`, which shouldn't be allowed.

Then it widened the return type of `func` to `Promise<void> | null`.

Reply:
@HenriBeck It affects the others as well, but only with `null` though:

<img width="364" alt="image" src="https://user-images.githubusercontent.com/52437/55459262-631e8500-55ef-11e9-84ce-3b9d34ba696a.png">

I'll fix these as well.

Reply:
Now the errors are in the right place:
<img width="265" alt="image" src="https://user-images.githubusercontent.com/52437/55459423-d0cab100-55ef-11e9-9f0c-009f2d31091e.png">

(With shorter typings they would mark the `func` as being invalid, not the `value`)

Reply:
Do you know why only `null` wasn't working? 🤔

Reply:
No, I think it's a TypeScript issue. I'll add a TODO for me to extract a repo and report it.

TBH I wouldn't have been surprised if `string` would also cause widening.

Not sure if they now give equal weight to all "consumers" of a generic type, or if the rightmost should still win.

Reply:
Just checked, this didn't change and is already "broken" in TS 3.3.

![image](https://user-images.githubusercontent.com/52437/55471325-c2898e80-5609-11e9-91ac-e2dd2d02b3e3.png)

I think this is caused by this issue from 2014. I was irritated by the `null` in the return type, but that was before `strict` mode, so…
https://github.com/Microsoft/TypeScript/issues/241